File: 1/2021.12.14/630.Course Schedule III/java/solution.java
class Solution {
    public int scheduleCourse(int[][] courses) {
        Arrays.sort(courses, (a, b) -> a[1] - b[1]);
        PriorityQueue<Integer> heap = new PriorityQueue<>((a, b) -> b - a);
        int total = 0;
        for (int[] course : courses) {
            int ti = course[0], di = course[1];
            if (total + ti <= di) {
                total += ti;
                heap.offer(ti);
            } else if (!heap.isEmpty() && heap.peek() > ti) {
                total -= heap.poll();
                total += ti;
                heap.offer(ti);
            }
        }
        return heap.size();
    }
}

File: 1/2021.12.14/630.Course Schedule III/python/solution.py
class Solution:
    def scheduleCourse(self, courses: List[List[int]]) -> int:
        courses.sort(key=lambda x : x[1])
        heap, total = [], 0
        for ti, di in courses:
            if total + ti <= di:
                total += ti
                heappush(heap, -ti)
            elif heap and -heap[0] > ti:
                total += heappop(heap)
                total += ti
                heappush(heap, -ti)
        return len(heap)
